example/EmbeddedChessPieces/src/ECPChessLogic/ECPChessPiece.cpp

```diff
@@ -1,19 +1,149 @@
 #include "ECPChessPiece.h"
 
-ECPChessPiece::ECPChessPiece(Dezibot &d, ECPChessField initialField, bool isWhite)
-    : dezibot(d), currentField(initialField), isWhite(isWhite) {};
+ECPChessPiece::ECPChessPiece(
+    Dezibot &d,
+    ECPMovement &ecpMovement,
+    ECPChessField initialField,
+    bool isWhite
+) : dezibot(d),
+    ecpMovement(ecpMovement),
+    currentField(initialField),
+    isWhite(isWhite),
+    currentDirection(isWhite ? NORTH : SOUTH) {};
 
 bool ECPChessPiece::move(ECPChessField newField) {
-    if (!isMoveValid(newField)) {
+    // show light depending on validity of requested movement
+    if (isMoveValid(newField)) {
+        setGreenLight(true);
+        delay(COLOR_DELAY);
+        setGreenLight(false);
+    } else {
+        setRedLight(true);
+        delay(COLOR_DELAY);
+        setRedLight(false);
         return false;
     }
 
-    // TODO: #4 implement actual movement on board
+    const int colDiff = (int) currentField.column - (int) newField.column;
+    const int rowDiff = (int) currentField.row - (int) newField.row;
+
+    if (colDiff != 0) {
+        moveHorizontally(colDiff);
+    }
+
+    if (rowDiff != 0) {
+        moveVertically(rowDiff);
+    }
 
     currentField = newField;
+    turnBackToInitialDirection();
+
     return true;
 };
 
 ECPChessField ECPChessPiece::getCurrentField() {
     return currentField;
 };
+
+void ECPChessPiece::moveHorizontally(int fieldsToMove) {
+    const Direction newDirection = fieldsToMove > 0 ? WEST : EAST;
+
+    // turn in the right direction if necessary
+    switch (currentDirection) {
+        case NORTH:
+            newDirection == WEST ? ecpMovement.turnLeft() : ecpMovement.turnRight();
+            break;
+        case EAST:
+            if (newDirection == WEST) {
+                ecpMovement.turnLeft();
+                ecpMovement.turnLeft();
+            }
+            break;
+        case SOUTH:
+            newDirection == WEST ? ecpMovement.turnRight() : ecpMovement.turnLeft();
+            break;
+        case WEST:
+            if (newDirection == EAST) {
+                // turn around
```

Comment:
```suggestion
            if (newDirection != WEST) {
```
Maybe this is easier to read and its consistent to `moveVertically`.

Reply:
Addressed in ee3da03 and e7879f8